### PR TITLE
[T5, generation] Add decoder caching for T5

### LINKED ISSUE #3605: 🐛 Summarization pipeline : T5-base much slower than BART-large
# 🐛 Bug

## Information

Model : `bart-large-cnn` and `t5-base`

Language : English

The problem arises when using : [this colab notebook](https://colab.research.google.com/drive/1iAIFX1QQiFm1F01vMmnAgFh4oH1H-K8W), using both BART and T5 with pipeline for Summarization.

Dataset : CNN/DM

## To reproduce

Run the notebook and measure time for inference between the 2 models. On my run, I have :

```
BART = 73s
T5 = 369s
```

## Expected behavior

I expected T5 to be at least as fast as BART, since there is less parameters (for the base version at least). Instead it takes much longer with T5...

@patrickvonplaten Do you happen to know why T5 is so slow ?

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This PR greatly speeds up the autoregressive decoding for T5 by storing past key / value states.
The summarization test: https://github.com/huggingface/transformers/blob/500aa12318ce5acd289d5edb6cb8266b3c3b162e/tests/test_modeling_t5.py#L260 now takes only 44s whereas before it took 311s -> 7.5x Speed up

This will also significantly speed up the translation and summarization pipelines when using T5.

- [x] Add key value state caching
- [x] Test for equal output on hard-coded tests
- [x] Add simple past tests including using an attention mask
- [x] update the docstring
- [x] clean up code

The caching design was already somewhat outcommented in place. It was cleaned, made functional and implemented very similar to GPT2's one.

### IMPORTANT: 
This PR has a breaking change, in that it increases the default output length of T5Model and T5ForConditionalGeneration from 4 to 5 (including the `past_key_value_states`).

### Future PR:
- [ ] Do the same for TF if this PR is accepted.

Would be nice if you could take a look @craffel @thomwolf @LysandreJik @sshleifer

### CODE REVIEW

File: src/transformers/modeling_t5.py [hunk 1]
```diff
@@ -421,31 +478,63 @@ def forward(
...
         head_mask=None,
+        past_key_value_state=None,
     ):
+
+        if self.is_decoder and self.output_past and past_key_value_state is not None:
+            assert (
+                len(past_key_value_state) == 4
+            ), "The should be 4 past states. 2 (past / key) for self attention. 2 (past / key) for cross attention. Got {} past key / value states".format(
+                len(past_key_value_state)
+            )
```
Comment (patrickvonplaten) [inline src/transformers/modeling_t5.py:489]: We store both the key value vectors of both the self attn layer and the cross attn layer so that we have 4 stored states per layer.

File: src/transformers/modeling_t5.py [hunk 2]
```diff
@@ -837,16 +968,29 @@ def forward(
 
         hidden_states = encoder_outputs[0]
 
+        # If decoding with past key value states, only the last tokens
+        # should be given as an input
+        if decoder_past_key_value_states is not None and self.decoder.output_past is True:
+            if decoder_input_ids is not None:
+                decoder_input_ids = decoder_input_ids[:, -1:]
+            if decoder_inputs_embeds is not None:
+                decoder_inputs_embeds = decoder_inputs_embeds[:, -1:]
```
Comment (patrickvonplaten) [inline src/transformers/modeling_t5.py:977]: That's a rather import design choice I made and which is different to GPT2. 
Even when the user is using key value caching, it is quite natural to still input all the input_ids (from `1` to `sequence_length` instead of just the last one as is done in GPT2). So using these lines of code it works both on inputting only the last `input_ids` and all of them.

File: src/transformers/modeling_t5.py [hunk 3]
```diff
@@ -948,16 +1104,34 @@ def forward(
...
             encoder_hidden_states=hidden_states,
             encoder_attention_mask=attention_mask,
             head_mask=head_mask,
         )
 
+        # insert decoder past at right place
+        # to speed up decoding
+        if self.decoder.output_past:
+            past = ((encoder_outputs, decoder_outputs[1]),)
+            decoder_outputs = decoder_outputs[:1] + past + decoder_outputs[2:]
```
Comment (patrickvonplaten) [inline src/transformers/modeling_t5.py:1133]: That's also important. As in done in Bart the past (`output[1]` in the generate fn) contains both the `encoder_outputs` and the cached key value states. This allows us to only work with the `past` variable in all `prepare_inputs_for_generation` functions, but it requires to have two pointers to the encoder outputs in all outputs for T5.

**Review by sshleifer — APPROVED** [review]
LGTM, epic speedup!
I don't like`config.output_past` but fine either way.

File: src/transformers/modeling_t5.py [hunk 4]
```diff
@@ -780,6 +897,17 @@ def set_input_embeddings(self, new_embeddings):
...
 
+    def set_output_past(self, do_output_past: bool):
+        self.config.output_past = do_output_past
+        self.decoder.output_past = do_output_past
+        self.encoder.output_past = do_output_past
+
+    def get_encoder(self):
+        return self.encoder
+
+    def get_decoder(self):
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/transformers/modeling_t5.py:908]: yeah in tests

File: src/transformers/modeling_t5.py [hunk 5]
```diff
@@ -968,27 +1143,49 @@ def forward(
...
+        past = (past[0],)
+        reordered_decoder_past = ()
+        for layer_past_states in decoder_past:
+            # get the correct batch idx from layer past batch dim
+            # batch dim of `past` is at 2nd position
+            reordered_layer_past_states = ()
+            for layer_past_state in layer_past_states:
+                # need to set correct `past` for each of the four key / value states
+                reordered_layer_past_states = reordered_layer_past_states + (
+                    torch.cat([layer_past_state[i].unsqueeze(0).clone() for i in beam_idx], dim=0),
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/transformers/modeling_t5.py]: super useful - don't know why I didn't use that before :D

File: src/transformers/modeling_t5.py (hunk 2 above)
Comment (LysandreJik) [reply src/transformers/modeling_t5.py:978]: This seems like a good idea to me, but that would definitely need to be referenced somewhere.

I think eventually we would like to do the same to GPT-2, in order to have a consistent API.

File: tests/test_modeling_t5.py [hunk 6]
```diff
@@ -167,24 +167,34 @@ def create_and_check_t5_model(
...
+            decoder_output, decoder_past, encoder_output = model(
                 input_ids=input_ids,
                 decoder_input_ids=decoder_input_ids,
                 attention_mask=attention_mask,
                 decoder_attention_mask=decoder_attention_mask,
             )
-            decoder_output, encoder_output = model(input_ids=input_ids, decoder_input_ids=decoder_input_ids)
+            decoder_output, decoder_past, encoder_output = model(
+                input_ids=input_ids, decoder_input_ids=decoder_input_ids
+            )
```
Comment (LysandreJik) [inline tests/test_modeling_t5.py:178]: Does the `decoder_past` get output by default?

**Review by LysandreJik — APPROVED** [review]
LGTM!

**Review by thomwolf — APPROVED** [review]
Look good to me

Just don't forget to update the PR description if the output of T5 with default settings has changed (see my last comment).

File: src/transformers/modeling_t5.py [hunk 7]
```diff
@@ -780,6 +897,17 @@ def set_input_embeddings(self, new_embeddings):
         self.encoder.set_input_embeddings(new_embeddings)
         self.decoder.set_input_embeddings(new_embeddings)
 
+    def set_output_past(self, do_output_past: bool):
+        self.config.output_past = do_output_past
+        self.decoder.output_past = do_output_past
```
Comment (yjernite) [inline src/transformers/modeling_t5.py:902]: You also need to set output_past in the blocks and attention layers:
`for block in self.decoder.block:
            block.output_past = do_output_past
            block.layer[0].SelfAttention.output_past = do_output_past
            block.layer[1].EncDecAttention.output_past = do_output_past`

File: src/transformers/modeling_t5.py [hunk 8]
```diff
@@ -872,6 +1017,11 @@ def __init__(self, config):
     def get_input_embeddings(self):
         return self.shared
 
+    def set_output_past(self, do_output_past: bool):
+        self.config.output_past = do_output_past
+        self.decoder.output_past = do_output_past
+        self.encoder.output_past = do_output_past
```
Comment (yjernite) [inline src/transformers/modeling_t5.py:1023]: Same comment as in the parent class, the output_past variables in the decoder blocks and attention layer need to be set.

Comment (patrickvonplaten) [reply src/transformers/modeling_t5.py]: Thanks for catching that @yjernite! Looking at this function this also makes me wonder whether it might be better to use a `use_cache` boolean flag argument instead of a `config.output_past` flag (which @sshleifer proposed and which is implemented in https://github.com/patrickvonplaten/transformers/blob/e20a76f52a40300357b6e1f70a3e896d39840095/src/transformers/modeling_bart.py#L414).

I guess the Pro's and Con's are:

Pros for `use_cache` / Cons for `output_past`:
- more control for the user as he decides at function call whether to output past or not
- seems logical to be correct to use it here, because it is not really an attribute of the model
- some models can't output the past or it doesn't make sense (models that don't use a causal mask)
- less under-the-hood workings and easier to see for the user what's going on

Cons for `use_cache` / Pros for `output_past`:
- we would have to change a lot of code in GPT2, GPT, CTRL, T5 and their docstrings
- might be a breaking change if we remove the `output_past` from the configs

I'm in favor of using `use_cache` as @sshleifer is I think. What is your opinion on that @LysandreJik, @thomwolf, @yjernite, @sshleifer?

Comment (patrickvonplaten) [reply src/transformers/modeling_t5.py]: @sshleifer and I could do the changes in a separate PR and we could default the `use_cache` to True so there shouldn't be breaking changes.

File: src/transformers/modeling_t5.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_t5.py:981]: Noted down to do a PR for GPT2 to implement that as well :-)

File: src/transformers/modeling_t5.py (hunk 8 above)
Comment (yjernite) [reply src/transformers/modeling_t5.py]: Definitely for. I had something similar in my local branch for a bit. Mostly I think it's safer / will lead to fewer silent bugs.

Comment (patrickvonplaten) [reply src/transformers/modeling_t5.py]: Also one thing to consider is that some memory efficient models would like to switch `use_cache` off to be able to run on long sequences, so it's not always better to use the cache.

File: src/transformers/modeling_t5.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_t5.py:983]: And updated the docstring that should clearly describe the behavior now.

File: src/transformers/modeling_t5.py (hunk 8 above)
Comment (thomwolf) [reply src/transformers/modeling_t5.py]: I'm ok with using `use_cache`. I just don't want this to be a breaking change if possible.
